src/freenas/usr/local/sbin/migrate_warden.py

```diff
@@ -0,0 +1,400 @@
+        iocage_root.umount()
+        iocage_root.delete()
+
+        try:
+            warden_dataset.snapshot(
+                f"{warden_dataset.name}@WardenMigration_{date}",
+                recursive=True)
+        except libzfs.ZFSException as e:
+            if e.code == libzfs.Error.EXISTS:
+                pass  # Snapshot exists.
+            else:
+                raise
+
+        try:
+            warden_dataset.send(
+                send_fd,
+                fromname=None,
+                toname=fromsnap,
+                flags={
+                    None if not self.verbose else libzfs.SendFlag.PROGRESS,
+                    libzfs.SendFlag.PROPS
+                }
+            )
+        finally:
+            try:
+                os.close(send_fd)
+            except OSError:
+                pass
+
+    def recv_dataset(self, recv_fd, dataset):
+        # Defining these here instead of directly giving it in the function call
+        # as that aids in readability
+        force = True
+        nomnt = False
+        try:
+            self.zfs.receive(dataset, recv_fd, force, nomnt)
+        finally:
+            try:
+                os.close(recv_fd)
+            except OSError:
+                pass
+
+
+class Migrate(object):
+    def __init__(self, jail, _dir, pool, verbose, loop):
+        self.jail = jail
+        self.dir = _dir
+        self.meta = f"{self.dir}/.{self.jail}.meta"
+        self.pool = pool
+        self.dataset = f"{self.dir}/{self.jail}"
+        self.zfs = libzfs.ZFS(history=True,
+                              history_prefix="<warden-migration>")
+        self.ZFS = ZFS(self.pool, self.dataset, self.zfs, verbose)
+        self.thread_pool_executor = ThreadPoolExecutor(2)
+        self.loop = loop
+        self.r_pipe, self.s_pipe = os.pipe()
+        self.iocage_root = f"{self.pool}/iocage/jails/{self.jail}/root"
+        self.warden_dataset = self.zfs.get_dataset_by_path(self.dataset)
+        self.date = str(datetime.datetime.utcnow()).split()[0]
+
+    async def migrate_jail(self):
+        files = {
+            "name":        "host",
+            "ip4_addr":    "ipv4",
+            "ip6_addr":    "ipv6",
+            "mac":         "mac",
+            "allow_props": "jail-flags",
+            "warden_id":   "id",
+            "vnet":        "vnet",
+            "boot":        "autostart",
+            "nat":         "nat"
+        }
+
+        pool_exists = self.ZFS.pool_exists()
+
+        if not pool_exists:
+            print(f"\nZFS Pool {self.pool} does not exist!\n"
+                  "Please supply a valid pool for iocage usage.")
+            exit(1)
+
+        migrated = self.ZFS.jail_exists()
+
+        if migrated:
+            print(f"  {self.jail} already exists in iocage, please destroy it"
+                  " first.")
+            return
+
+        props = {}
+
+        for ioc_prop, warden_prop in files.items():
+            prop = self.jail_props(warden_prop)
+
```

Comment:
Agreed, this will have to be supported in some form. Brandon, can you add ip4_addr=dhcp to iocage and use that functionality?